modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java

```diff
@@ -0,0 +1,522 @@
+                );
+
+                raftServers.put(addr, raftServer);
+
+                afterNodeStart(localPeer.consistentId(), cluster, dataPath, peersAndLearners, eventsClientListener);
+            }
+
+            if (addr.port() == clientPort) {
+                assertTrue(isServerAddress.test(addr));
+
+                raftClient = startTopologyAwareClient(cluster, clusterServices, isServerAddress, nodes, eventsClientListener, true);
+            }
+        }
+
+        return raftClient;
+    }
+
+    private TopologyAwareRaftGroupService startTopologyAwareClient(
+            ClusterService localClusterService,
+            Map<NetworkAddress, ClusterService> clusterServices,
+            Predicate<NetworkAddress> isServerAddress,
+            int nodes,
+            RaftGroupEventsClientListener eventsClientListener,
+            boolean notifyOnSubscription
+    ) {
+        if (eventsClientListener == null) {
+            eventsClientListener = new RaftGroupEventsClientListener();
+
+            var finalEventsClientListener = eventsClientListener;
+            localClusterService.messagingService().addMessageHandler(RaftMessageGroup.class, (msg, sender, correlationId) -> {
+                if (msg instanceof LeaderChangeNotification) {
+                    LeaderChangeNotification msg0 = (LeaderChangeNotification) msg;
+
+                    ClusterNode node = localClusterService.topologyService().getByConsistentId(sender);
+                    finalEventsClientListener.onLeaderElected(msg0.groupId(), node, msg0.term());
+                }
+            });
+        }
+
+        var commandsMarshaller = new ThreadLocalOptimizedMarshaller(localClusterService.serializationRegistry());
+
+        return TopologyAwareRaftGroupService.start(
+                GROUP_ID,
+                localClusterService,
+                FACTORY,
+                raftConfiguration,
+                peersAndLearners(clusterServices, isServerAddress, nodes),
+                true,
+                executor,
+                new LogicalTopologyServiceTestImpl(localClusterService),
+                eventsClientListener,
+                notifyOnSubscription,
+                commandsMarshaller
+        ).join();
+    }
+
+    private static PeersAndLearners peersAndLearners(
+            Map<NetworkAddress, ClusterService> clusterServices,
+            Predicate<NetworkAddress> isServerAddress,
+            int nodes
+    ) {
+        return PeersAndLearners.fromConsistentIds(
+                getNetworkAddresses(nodes).stream().filter(isServerAddress)
+                        .map(netAddr -> clusterServices.get(netAddr).topologyService().localMember().name()).collect(
+                                toSet()));
+    }
+
+    /**
+     * Generates a node address for each node.
+     *
+     * @param nodes Node count.
+     * @return List on network addresses.
+     */
+    private static List<NetworkAddress> getNetworkAddresses(int nodes) {
```

Comment:
you could use `org.apache.ignite.internal.network.utils.ClusterServiceTestUtils#findLocalAddresses`, up to you